starter-code/routes/index.js

```diff
@@ -1,9 +1,76 @@
-var express = require('express');
-var router = express.Router();
+const express        = require("express");
+const router         = express.Router();
+// User model
+const User           = require("../models/user");
+// Bcrypt to encrypt passwords
+const bcrypt         = require("bcrypt");
+const saltRounds   = 10;
+const ensureLogin = require("connect-ensure-login");
+const passport      = require("passport");
 
-/* GET home page. */
-router.get('/', function(req, res, next) {
-  res.render('index', { title: 'Express' });
+
+router.get('/', ensureNotAuthenticated, (req, res, next) => {
+  res.render('passport/signup');
+});
+
+router.post('/signup', ensureNotAuthenticated, (req, res, next) => {
+
+      const newUser = User({
+          username: req.body.username,
+          password: bcrypt.hashSync(req.body.password, bcrypt.genSaltSync(saltRounds))
+      });
+  
+      if (req.body.username === "" || req.body.password === "") {
+          res.render("passport/signup", {
+            errorMessage: "All fields required to sign-up"
+          });
+          return;
+      }
+  
+      User.findOne({ "username": newUser.username }, "username", (err, user) => {
+          if (user !== null) {
+            res.render("passport/signup", {
+              errorMessage: "That username already exists"
+            });
+            return;
+          }
+      })
+  
+     newUser.save((err) => {
```

Comment:
WARNING!!!

This piece of code is going to give you an error when you try to sign up with a username that already exists. That is because, no matter what happens after the User.findOne... the app is going to try to save the user. Then, because it renders a page if the user exists and it tries to render (or redirect) again after saving the user, it throws an error. Pretty bad, eh? Prepare yourself because there is more. You are saving the user either it exists or not. Yes, you are duplicating users with the same username. :scream:

You can avoid this nightmare saving inside the User.findOne... either with an else after the if or just putting the code after the if (you can do that because you use the return).  I prefer the else.